addons/mrp/models/mrp_production.py

```diff
@@ -1755,11 +1755,12 @@ def button_mark_done(self):
             return res
 
         if self.env.context.get('mo_ids_to_backorder'):
-            productions_to_backorder = self.browse(self.env.context['mo_ids_to_backorder'])
-            productions_not_to_backorder = self - productions_to_backorder
+            productions = self.browse(self.env.context['mo_ids_to_backorder'])
+            productions_to_backorder = productions.filtered(lambda p: p.picking_type_id.create_backorder != 'never')
+            productions_not_to_backorder = self + productions - productions_to_backorder
         else:
-            productions_not_to_backorder = self
-            productions_to_backorder = self.env['mrp.production']
+            productions_to_backorder = self.filtered(lambda p: p.picking_type_id.create_backorder == 'always')
+            productions_not_to_backorder = self - productions_to_backorder
```

Comment:
This is incorrect. The MOs should already be excluded from the `mo_ids_to_backorder` so you shouldn't need to check for them again here. I.e. you should add the check in `mrp_production_backorder`